### DISCUSSION TITLE
Routing checks handlers in each request

### DISCUSSION THREAD

**Discussion by sk-** [encode/starlette#2540] [Potential Issue] [answered]
The logic in `routing.py` checks whether the handler is async in each request.

https://github.com/encode/starlette/blob/554f368809e0d891a699667faf0cfbb20057eeb2/starlette/routing.py#L71

This does not seem to be required and if we do the check only once at the beginning of `request_response`, we achieve some performance gains as seen in https://github.com/encode/starlette/pull/2536/files

**Comment by Kludex** [comment, marked as answer]
- Worked on https://github.com/encode/starlette/pull/2536. Thanks!
